lib/Sema/CSApply.cpp

```diff
@@ -3988,6 +3988,20 @@ namespace {
       if (E->getType()->hasError())
         return E;
 
+      // If a component is already resolved, then all of them should be
+      // resolved, and we can let the expression be. This might happen when
+      // re-checking a failed system for diagnostics.
+      if (!E->getComponents().empty()
+          && E->getComponents().front().isResolved()) {
+        assert([&]{
+          for (auto &c : E->getComponents())
+            if (!c.isResolved())
+              return false;
+          return true;
+        }());
```

Comment:
Hm. To me the immediately-called lambda is already weird enough to counter that, but okay.

```c++
assert(llvm::all_of(E->getCompanents(),
                    [](const KeyPathExpr::Component &c) {
  return c.isResolved();
}));
```